pallas-applying/src/byron.rs

```diff
@@ -38,3 +58,183 @@ fn check_ins_in_utxos(tx: &Tx, utxos: &UTxOs) -> ValidationResult {
     }
     Ok(())
 }
+
+fn check_outs_have_lovelace(tx: &Tx) -> ValidationResult {
+    for output in tx.outputs.iter() {
+        if output.amount == 0 {
+            return Err(ValidationError::OutputWithoutLovelace);
+        }
+    }
+    Ok(())
+}
+
+fn check_fees(tx: &Tx, size: &u64, utxos: &UTxOs, prot_pps: &ByronProtParams) -> ValidationResult {
+    let mut inputs_balance: u64 = 0;
+    for input in tx.inputs.iter() {
+        match utxos
+            .get(&MultiEraInput::from_byron(input))
+            .and_then(MultiEraOutput::as_byron)
+        {
+            Some(byron_utxo) => inputs_balance += byron_utxo.amount,
+            None => return Err(ValidationError::UnableToComputeFees),
+        }
+    }
+    let mut outputs_balance: u64 = 0;
+    for output in tx.outputs.iter() {
+        outputs_balance += output.amount
+    }
+    let total_balance: u64 = inputs_balance - outputs_balance;
+    let min_fees: u64 = prot_pps.min_fees_const + prot_pps.min_fees_factor * size;
+    if total_balance < min_fees {
+        return Err(ValidationError::FeesBelowMin);
+    }
+    Ok(())
+}
+
+fn check_size(size: &u64, prot_pps: &ByronProtParams) -> ValidationResult {
+    if *size > prot_pps.max_tx_size {
+        return Err(ValidationError::MaxTxSizeExceeded);
+    }
+    Ok(())
+}
+
+fn get_tx_size(tx: &Tx) -> Result<u64, ValidationError> {
+    let mut buff: Vec<u8> = Vec::new();
+    match encode(tx, &mut buff) {
+        Ok(()) => Ok(buff.len() as u64),
+        Err(_) => Err(ValidationError::UnknownTxSize),
+    }
+}
+
+pub enum TaggedSignature<'a> {
+    PkWitness(&'a Signature),
+    RedeemWitness(&'a Signature),
+}
+
+impl<'a> TaggedSignature<'a> {
+    fn get_raw_signature(&'a self) -> &'a Signature {
+        match self {
+            TaggedSignature::PkWitness(sign) => sign,
+            TaggedSignature::RedeemWitness(sign) => sign,
+        }
+    }
+}
+
+fn check_witnesses(mtxp: &MintedTxPayload, utxos: &UTxOs, prot_magic: &u32) -> ValidationResult {
+    let tx: &Tx = &mtxp.transaction;
+    let tx_hash: Hash<32> = mtxp.transaction.original_hash();
+    let witnesses: Vec<(&PubKey, TaggedSignature)> = tag_witnesses(&mtxp.witness)?;
+    let tx_inputs: &Vec<TxIn> = &tx.inputs;
+    for input in tx_inputs {
+        let tx_out: &TxOut = find_tx_out(input, utxos)?;
+        let (pub_key, sign): (&PubKey, &TaggedSignature) = find_witness(tx_out, &witnesses)?;
+        let signature: &Vec<u8> = sign.get_raw_signature();
+        let mut se = Serializer::new_vec();
+        serialize_tag(&mut se, sign)?;
+        se.serialize(&prot_magic)
+            .map_err(|_| ValidationError::UnableToProcessWitnesses)?;
+        se.serialize(&tx_hash.as_ref())
+            .map_err(|_| ValidationError::UnableToProcessWitnesses)?;
+        let data_to_verify: Vec<u8> = se.finalize();
+        if !ed25519::verify(&data_to_verify, &pub_key.as_slice()[0..32], signature) {
```

Comment:
we already have a function in `pallas::crypto::keys` to verify ed25519. We should use that one.